REGRESSION (iPadOS 18): Right-to-left text pasted from Safari into Pages is left aligned

whsieh opened the pull request:
#### bc0fb06c1c57d1718e0e663baa4f99130afb07a6
<pre>
REGRESSION (iPadOS 18): Right-to-left text pasted from Safari into Pages is left aligned
<a href="https://bugs.webkit.org/show_bug.cgi?id=286404">https://bugs.webkit.org/show_bug.cgi?id=286404</a>
<a href="https://rdar.apple.com/139029945">rdar://139029945</a>

Reviewed by Abrar Rahman Protyasha.

In iOS 17.4 and iPadOS 18, copying text within a single RTL paragraph in Safari and pasting into
Pages results in LTR text being inserted. This is because BrowserEngineKit adoption
(`async_text_input_i(phone|pad)`) is enabled since these updates, which causes WebKit to stop
directly converting selected text content into attributed string data upon copying; instead, we now
just write web archive and HTML data to the system pasteboard, which is lazily converted into
`NSAttributedString` by the destination app, if needed. This conversion relies on UIFoundation,
which (through `nsattributedstringagent`) loads a web view using the web archive data, and then uses
WebKit APIs to obtain an `NSAttributedString` from the web view.

Previously, WebKit would&apos;ve written an attributed string directly to the pasteboard, where the
attributes of the text contain an `NSParagraphStyle` that has RTL base writing direction; Pages
would then honor this by inserting an RTL paragraph. However, after enabling lazy attributed string
conversion, the resulting attributed string (derived from web archive data) no longer has a base
writing direction set. That&apos;s because the web archive markup contains only a single inline `span`
element which doesn&apos;t have any `direction` set, since the selection doesn&apos;t (fully) encompass any
block-level elements, which means any information about paragraph writing direction is lost upon
copying:

=&gt; Copied markup: `&lt;head&gt;&lt;meta charset=&apos;utf-8&apos;&gt;&lt;/head&gt;&lt;span&gt;كتروني هو&lt;/span&gt;`

While we could fix this by wrapping the web archive markup with a block-level `div` or `p` element
that carries over the `direction`, this could cause us to paste an extra paragraph, when the copied
content should really be treated as inline content (added to an existing paragraph):

=&gt; `&lt;head&gt;&lt;meta charset=&apos;utf-8&apos;&gt;&lt;/head&gt;&lt;div dir=&apos;rtl&apos;&gt;كتروني هو&lt;/div&gt;`

Instead, this patch addresses the issue by wrapping the web archive data in a `body` element. During
sanitization, instead of implicitly generating a `body` element, we&apos;ll parse the `body` element from
markup, along with the `dir=&apos;rtl&apos;` attribute. The contents of the body then properly contain RTL
text, which allows attributed string conversion to produce paragraph styles with an RTL base writing
direction set.

=&gt; `&lt;head&gt;&lt;meta charset=&apos;utf-8&apos;&gt;&lt;/head&gt;&lt;body dir=&apos;rtl&apos;&gt;&lt;span&gt;كتروني هو&lt;/span&gt;&lt;/body&gt;`

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-drop-rich-text-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-rich-text-expected.txt:
* LayoutTests/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/platform/ios/editing/pasteboard/onpaste-text-html-expected.txt:
* LayoutTests/platform/mac/editing/pasteboard/onpaste-text-html-expected.txt:
* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* Source/WebCore/editing/markup.cpp:

Rebaseline several layout tests.

* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* Source/WebCore/editing/markup.cpp:
(WebCore::serializePreservingVisualAppearanceInternal):

See description above.

(WebCore::serializePreservingVisualAppearance):
(WebCore::sanitizedMarkupForFragmentInDocument):
* Source/WebCore/editing/markup.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createFromSelection):

Pass in `PreserveDirectionForInlineText::Yes`.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/rtl-bidi-text.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyHTML.mm:
(copyAndLoadAttributedStringUsingWebArchive):
(TEST(CopyHTML, SanitizationPreservesRelativeURLInAttributedString)):
(TEST(CopyHTML, CopyingRightToLeftTextPreservesDirection)):

Add an API test to exercise the fix, by:
- Copying selected RTL text.
- Using the copied web archive to generate an attributed string.
- Verifying that the attributed string has a RTL base writing direction.

Canonical link: <a href="https://commits.webkit.org/289323@main">https://commits.webkit.org/289323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5837a07cf8edee35935214c230dc658c0e3584a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86418 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91326 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37215 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14004 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66907 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24690 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78271 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47227 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4528 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32595 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36323 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75052 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33469 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93164 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13618 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9876 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75698 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13825 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74880 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18430 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19146 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17536 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13641 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18918 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13394 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15178 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -358,6 +358,27 @@ class StyledMarkupAccumulator final : public MarkupAccumulator {
+
+        auto directionValue = renderer->writingMode().bidiDirection() == TextDirection::LTR ? "ltr"_s : "rtl"_s;
+        m_reversedPrecedingMarkup.append(makeString("<body dir=\""_s, WTFMove(directionValue), "\">"_s));
+        append("</body>"_s);
```
Nit either way, but why not just put `</body>` in the previous line?

whsieh replied:
Hm... do you mean something like this?

```
m_reversedPrecedingMarkup.append(makeString("<body dir=\""_s, WTFMove(directionValue), "\"></body>"_s));
```

...if so, that would produce an invalid markup string:

```
<body dir=…></body><span style=…></span>
```

aprotyas replied:
oh, my bad, I didn't pay attention to the markup.

whsieh replied:
No worries! Will keep this as-is. Thanks for the review!

whsieh committed:
REGRESSION (iPadOS 18): Right-to-left text pasted from Safari into Pages is left aligned
https://bugs.webkit.org/show_bug.cgi?id=286404
rdar://139029945

Reviewed by Abrar Rahman Protyasha.

In iOS 17.4 and iPadOS 18, copying text within a single RTL paragraph in Safari and pasting into
Pages results in LTR text being inserted. This is because BrowserEngineKit adoption
(`async_text_input_i(phone|pad)`) is enabled since these updates, which causes WebKit to stop
directly converting selected text content into attributed string data upon copying; instead, we now
just write web archive and HTML data to the system pasteboard, which is lazily converted into
`NSAttributedString` by the destination app, if needed. This conversion relies on UIFoundation,
which (through `nsattributedstringagent`) loads a web view using the web archive data, and then uses
WebKit APIs to obtain an `NSAttributedString` from the web view.

Previously, WebKit would've written an attributed string directly to the pasteboard, where the
attributes of the text contain an `NSParagraphStyle` that has RTL base writing direction; Pages
would then honor this by inserting an RTL paragraph. However, after enabling lazy attributed string
conversion, the resulting attributed string (derived from web archive data) no longer has a base
writing direction set. That's because the web archive markup contains only a single inline `span`
element which doesn't have any `direction` set, since the selection doesn't (fully) encompass any
block-level elements, which means any information about paragraph writing direction is lost upon
copying:

=> Copied markup: `<head><meta charset='utf-8'></head><span>كتروني هو</span>`

While we could fix this by wrapping the web archive markup with a block-level `div` or `p` element
that carries over the `direction`, this could cause us to paste an extra paragraph, when the copied
content should really be treated as inline content (added to an existing paragraph):

=> `<head><meta charset='utf-8'></head><div dir='rtl'>كتروني هو</div>`

Instead, this patch addresses the issue by wrapping the web archive data in a `body` element. During
sanitization, instead of implicitly generating a `body` element, we'll parse the `body` element from
markup, along with the `dir='rtl'` attribute. The contents of the body then properly contain RTL
text, which allows attributed string conversion to produce paragraph styles with an RTL base writing
direction set.

=> `<head><meta charset='utf-8'></head><body dir='rtl'><span>كتروني هو</span></body>`

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-drop-rich-text-expected.txt:
* LayoutTests/editing/pasteboard/data-transfer-get-data-on-paste-rich-text-expected.txt:
* LayoutTests/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/platform/ios/editing/pasteboard/onpaste-text-html-expected.txt:
* LayoutTests/platform/mac/editing/pasteboard/onpaste-text-html-expected.txt:
* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* Source/WebCore/editing/markup.cpp:

Rebaseline several layout tests.

* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* Source/WebCore/editing/markup.cpp:
(WebCore::serializePreservingVisualAppearanceInternal):

See description above.

(WebCore::serializePreservingVisualAppearance):
(WebCore::sanitizedMarkupForFragmentInDocument):
* Source/WebCore/editing/markup.h:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::createFromSelection):

Pass in `PreserveDirectionForInlineText::Yes`.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/rtl-bidi-text.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyHTML.mm:
(copyAndLoadAttributedStringUsingWebArchive):
(TEST(CopyHTML, SanitizationPreservesRelativeURLInAttributedString)):
(TEST(CopyHTML, CopyingRightToLeftTextPreservesDirection)):

Add an API test to exercise the fix, by:
- Copying selected RTL text.
- Using the copied web archive to generate an attributed string.
- Verifying that the attributed string has a RTL base writing direction.

Canonical link: https://commits.webkit.org/289323@main
